### ISSUE TITLE
HTMLTableElement.prototype.rows does not reflect reality until length called

### ISSUE THREAD

**Issue by moll** [jsdom/jsdom#742]
I wasn't able to reproduce this immediately in a clean test case to share, but I stumbled upon an issue where `table.deleteRow(0)` does not change `table.rows` indices until `table.rows.length` is accessed.

Took me a while to understand that this was happening. A kind of a heisenbug.

**Comment by levimm** [comment]
If I understand it right, the HTMLCollection maintains two arrays and when you modify the HTMLCollection, it modify only one of the arrays, and when you access, you are actually accessing another one. So it needs to call update to synchronize them. Am I right?
Could you give me some example why jsdom does in this way?

**Comment by domenic** [comment]
Should be fixed by https://github.com/tmpvar/jsdom/pull/1934, although we should add specific tests to confirm the caching isn't messing us up somehow.

**Comment by TimothyGu** [comment]
Yes it's fixed! Did have to change the test to account for automatic tbody adding though:

```js
const { JSDOM } = require(".");

const html = `
<table id="table">
  <tbody>
    <tr><td>Cell 1</td></tr>
    <tr><td>Cell 2</td></tr>
  </tbody>
</table>
<script>
  var table = document.getElementById('table');
  var tbody = table.children[0];

  console.log(table.rows[0].cells[0].textContent);

  var row1 = table.rows[0];
  tbody.removeChild(row1);
  tbody.appendChild(row1);

  console.log(table.rows[0].cells[0].textContent);

  table.rows.length;

  console.log(table.rows[0].cells[0].textContent);
</script>
`;

new JSDOM(html, { runScripts: 'dangerously' });
```

now outputs

```
Cell 1
Cell 2
Cell 2
```

as expected.
